mitodl/mitxpro
# refactor: standardize Catalog & API courseware

### What are the relevant tickets?
https://github.com/mitodl/hq/issues/5266

### Description (What does it do?)
- Standardize the logic between Catalog and courseware APIs
- Refactor the code to be more generic in handling external and internal courseware
- Add more tests and Refactor existing tests
- Removes unnecessary code




### How can this be tested?
**NOTE:** This PR would probably require extensive testing of Catalog and courseware APIs and smoke testing of the whole application.

- On master create a data set of various internal and external Courses & Programs
- Check the courses on Catalog and Courses in the `/api/courses`
- Verify the existing logic based on [this sheet](https://docs.google.com/spreadsheets/d/1VqWs7CEN7J47iEiwtReeAS6nyQlyTp9qTR9mg9_V4dI/edit?gid=0#gid=0). You can see that there is a difference between which courses are displayed on Catalog and which are part of the APIs.
- Create a scenario while being on master where you see difference of courses on Catalog and `api/courses`

**Shift to this branch**

- Verify that you see exactly the same set of courses/programs on catalog and the APIs
- Note that, The product is now required for all the courses and programs to be displayable on Catalog and APIs
- Note that, The course run start/enrollment end date conditions are applied on both 
- Perform filter checking, sort by on catalog
- Check all the APIs `/api/courses`, `/api/programs`, `/api/products`


### Additional Context
